### PR TITLE
feat(Locomotion): Integrate Immersive Movement (Now Real Movement) into VRTK

### PR DESCRIPTION
**Pull request by Highsight**
Adds ability to move the play area through the measurement of arm and/or headset Y movements.  I set the Example scene to #42, as it seems we have some other PRs that have beaten me to #40.

### CODE REVIEW

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 34d5a7d95da1ef99314299d969a7a30a6e2e3332
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 46c1b0029c0f2be4ff3f87201e7a5b10469dcc76
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: fe07543e456b89d0dac9090105591710ea5796cd
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/thestonefox/SteamVR_Unity_Toolkit

---
This message was auto-generated by https://gitcop.com

**Comment by thestonefox** [comment]
I'd recommend squashing the commits into 1 commit.

also the gitcop message is complaining about the git commit message which you'll need to change with a `git commit --amend` just changing the message on github doesn't change the git commit messages.

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 34d5a7d95da1ef99314299d969a7a30a6e2e3332
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 46c1b0029c0f2be4ff3f87201e7a5b10469dcc76
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: fe07543e456b89d0dac9090105591710ea5796cd
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 72c1f6056e0ea4f9a289af3da39748c76a42ded3
  * Your subject line is longer than 72 characters

Guidelines are available at https://github.com/thestonefox/SteamVR_Unity_Toolkit

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 34d5a7d95da1ef99314299d969a7a30a6e2e3332
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 46c1b0029c0f2be4ff3f87201e7a5b10469dcc76
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: fe07543e456b89d0dac9090105591710ea5796cd
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 72c1f6056e0ea4f9a289af3da39748c76a42ded3
  * Your subject line is longer than 72 characters

Guidelines are available at https://github.com/thestonefox/SteamVR_Unity_Toolkit

---
This message was auto-generated by https://gitcop.com

**Comment by Highsight** [comment]
I'm a total noob when it comes to git, but I got it down to 2 commits.  One is the one you're looking for (829b908) and the other is some sort of Merge branch that I was forced to do before I could commit the squashed code. (6cf2504)  I have no idea why GitCop is still complaining though.  Am I supposed to be able to actually remove those other commits from the history or something?

**Comment by thestonefox** [comment]
You'll need to rebase the remote upstream master into your branch to remove that merge commit (that merge commit shouldn't be there)

File: Assets/VRTK/Scripts/VRTK_RealMovement.cs [hunk 1]
```diff
@@ -0,0 +1,350 @@
+﻿// Touchpad Walking|Scripts|0140
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_RealMovement.cs]: This needs updating to the specific script

File: Assets/VRTK/Scripts/VRTK_RealMovement.cs [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+        /// <param name="ControllersOnly">Track only the controllers for movement calculations.</param>
+        /// <param name="HeadsetOnly">Track only headset for movement caluclations.</param>
+        public enum ControlOptions
+        {
+            HeadsetAndControllers,
+            ControllersOnly,
+            HeadsetOnly,
+        }
+        
+        public enum DirectionalMethod
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_RealMovement.cs]: Needs documentation

File: Assets/VRTK/Scripts/VRTK_RealMovement.cs [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+        {
+            Gaze,
+            DumbDecoupling,
+            SmartDecoupling
+        }
+
+        public bool LeftController
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_RealMovement.cs]: Needs docs

File: Assets/VRTK/Scripts/VRTK_RealMovement.cs [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+        {
+            get { return leftController; }
+            set
+            {
+                leftController = value;
+                SetControllerListeners(controllerLeftHand);
+            }
+        }
+
+        public bool RightController
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_RealMovement.cs]: Needs docs

File: Assets/VRTK/Scripts/VRTK_RealMovement.cs [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+        [Tooltip("If this is checked then the left controller touchpad will be enabled to move the play area. It can also be toggled at runtime.")]
+        [SerializeField]
+        private bool leftController = true;
+        [Tooltip("If this is checked then the right controller touchpad will be enabled to move the play area. It can also be toggled at runtime.")]
+        [SerializeField]
+        private bool rightController = true;
+
+        [Tooltip("Select which trackables are used to determine movement.")]
+        [SerializeField]
+        private ControlOptions m_controlOptions = ControlOptions.HeadsetAndControllers;
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_RealMovement.cs]: the style of the toolkit doesn't use snake case variables. only use camel case. e.g. `controlOptions`

File: Assets/VRTK/Scripts/VRTK_RealMovement.cs [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+
+        [SerializeField]
+        [Tooltip("Lower to decrease speed, raise to increase.")]
+        private float m_speedScale = 1;
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_RealMovement.cs]: only camel case variable names

File: Assets/VRTK/Scripts/VRTK_RealMovement.cs [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+
+        [SerializeField]
+        [Tooltip("The max speed you can move in game units. (If 0 or less, max speed is uncapped)")]
+        private float m_maxSpeed = 4;
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_RealMovement.cs]: only camel case variable names

File: Assets/VRTK/Scripts/VRTK_RealMovement.cs [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+
+        [SerializeField]
+        [Tooltip("How your movement direction will be determined.  The Gaze method tends to lead to the least motion sickness.  Smart decoupling is still a Work In Progress.")]
+        private DirectionalMethod m_dirMethod = DirectionalMethod.Gaze;
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_RealMovement.cs]: only camel case variable names

File: Assets/VRTK/Scripts/VRTK_RealMovement.cs [hunk 9]
```diff
@@ -0,0 +1,350 @@
...
+
+        [SerializeField]
+        [Tooltip("If we're using Smart Decoupling, all trackables must be within this degree threshold to change direction.")]
+        private float m_smartDecoupleThreshold = 30f;
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_RealMovement.cs]: only camel case variable names

Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_RealMovement.cs]: Also, why are all these variables private but serialized?

How would you change them at runtime?

File: Assets/VRTK/Scripts/VRTK_RealMovement.cs [hunk 10]
```diff
@@ -0,0 +1,350 @@
...
+        private Dictionary<Transform, Vector3> m_initialTrackableDirs = new Dictionary<Transform, Vector3>();
+
+        private float m_curSpeed = 0.00f;
+
+        private Vector3 m_direction = new Vector3();
+
+        private bool active = false;
+
+        // If we're in Unity 5.4, we need to make sure we are using the eyes, not the head.
+        void Awake()
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_RealMovement.cs]: all methods need the accessor level on them e.g. `private void Awake()`

Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_RealMovement.cs]: Also private methods should be listed after public and protected methods

File: Assets/VRTK/Scripts/VRTK_RealMovement.cs [hunk 11]
```diff
@@ -0,0 +1,350 @@
...
+        {
+            controllerLeftHand = VRTK_DeviceFinder.GetControllerLeftHand();
+            controllerRightHand = VRTK_DeviceFinder.GetControllerRightHand();
+            headset = VRTK_DeviceFinder.HeadsetTransform();
+
+            setControlOptions(m_controlOptions);
+
+            touchpadPressed = new ControllerInteractionEventHandler(DoTouchpadDown);
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_RealMovement.cs]: you should probably do things like this in the `OnEnable` then you can deregister events in the `OnDisable`

File: Assets/VRTK/Scripts/VRTK_RealMovement.cs [hunk 12]
```diff
@@ -0,0 +1,350 @@
...
+            touchpadUp = new ControllerInteractionEventHandler(DoTouchpadUp);
+        }
+
+        public void setControlOptions(ControlOptions p_controlOptions)
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_RealMovement.cs]: Method names should start capitalised `SetControlOptions`

File: Assets/VRTK/Scripts/VRTK_RealMovement.cs [hunk 13]
```diff
@@ -0,0 +1,350 @@
...
+            }
+
+            if (m_controlOptions.Equals(ControlOptions.HeadsetAndControllers) || m_controlOptions.Equals(ControlOptions.HeadsetOnly))
+            {
+                m_trackedObjects.Add(headset.transform);
+            }
+        }
+
+        // Use this for initialization
+        void Start()
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_RealMovement.cs]: `private void Start()`

File: Assets/VRTK/Scripts/VRTK_RealMovement.cs [hunk 14]
```diff
@@ -0,0 +1,350 @@
...
+                m_trackedObjects.Add(controllerRightHand.transform);
+            }
+
+            if (m_controlOptions.Equals(ControlOptions.HeadsetAndControllers) || m_controlOptions.Equals(ControlOptions.HeadsetOnly))
+            {
+                m_trackedObjects.Add(headset.transform);
+            }
+        }
+
+        // Use this for initialization
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_RealMovement.cs]: Delete this default comment

File: Assets/VRTK/Scripts/VRTK_RealMovement.cs [hunk 15]
```diff
@@ -0,0 +1,350 @@
...
+            m_initalGaze = Vector3.zero;
+            m_initialTrackableDirs = new Dictionary<Transform, Vector3>();
+            m_direction = Vector3.zero;
+            m_curSpeed = 0;
+
+            active = false;
+        }
+
+        // FixedUpdate is called at a regular rate.
+        void FixedUpdate()
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_RealMovement.cs]: `private void FixedUpdate()`

File: Assets/VRTK/Scripts/VRTK_RealMovement.cs [hunk 16]
```diff
@@ -0,0 +1,350 @@
...
+                float listAverage = 0;
+
+                float m_playAreaDeltaY = Mathf.Abs(m_playAreaPreviousY - transform.position.y);
+
+                foreach (Transform trackedObj in m_trackedObjects)
+                {
+                    // Get the amount of Y movement that's occured since the last update.
+                    float deltaYPostion = Mathf.Abs(m_previousYPositions[trackedObj] - trackedObj.transform.localPosition.y);
+
+                    //deltaYPostion = Mathf.Max(0, deltaYPostion - (m_playAreaDeltaY));
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_RealMovement.cs]: Remove comments that aren't being used

File: Assets/VRTK/Scripts/VRTK_RealMovement.cs [hunk 17]
```diff
@@ -0,0 +1,350 @@
...
+
+                    // Convenience code.
+                    List<float> trackedObjList = m_movementList[trackedObj];
+
+                    if ((m_playerBody != null && (!m_playerBody.IsFalling() /*|| m_playerBody.IsClimbing()*/)) || m_playerBody == null)
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_RealMovement.cs]: Remove inline comment

File: Assets/VRTK/Scripts/VRTK_RealMovement.cs [hunk 18]
```diff
@@ -0,0 +1,350 @@
...
+                    float avg = sum / m_averagePeriod;
+
+                    // Add the average to the the list average.
+                    listAverage += avg;
+                }
+
+                float speed = ((m_speedScale * 350) * (listAverage / m_trackedObjects.Count));
+
+                if (speed > m_maxSpeed && m_maxSpeed >= 0)
+                    speed = m_maxSpeed;
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_RealMovement.cs]: Missing brackets

File: Assets/VRTK/Scripts/VRTK_RealMovement.cs [hunk 19]
```diff
@@ -0,0 +1,350 @@
...
+                        float curXDir = headset.rotation.eulerAngles.y;
+                        if (curXDir <= m_smartDecoupleThreshold)
+                        {
+                            curXDir += 360;
+                        }
+                    
+                        closeEnough = closeEnough && (Mathf.Abs(curXDir - controllerLeftHand.transform.rotation.eulerAngles.y) <= m_smartDecoupleThreshold);
+                        closeEnough = closeEnough && (Mathf.Abs(curXDir - controllerRightHand.transform.rotation.eulerAngles.y) <= m_smartDecoupleThreshold);
+
+                        if (closeEnough)
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_RealMovement.cs]: Missing brackets

File: Assets/VRTK/Scripts/VRTK_RealMovement.cs [hunk 20]
```diff
@@ -0,0 +1,350 @@
...
+
+            foreach (Transform trackedObj in m_trackedObjects)
+            {
+                // Get delta postions and rotations
+                m_previousYPositions[trackedObj] = trackedObj.transform.localPosition.y;
+            }
+
+            Vector3 movement = (m_direction * m_curSpeed) * Time.fixedDeltaTime;
+
+            /*if (m_playerBody != null)
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_RealMovement.cs]: remove commented out code

File: Assets/VRTK/Scripts/VRTK_RealMovement.cs [hunk 21]
```diff
@@ -0,0 +1,350 @@
...
+                movement = movement + m_playerBody.getImmersiveMovementModifier() - (cameraDistance);
+                m_playerBody.resetImmersiveMovementModifier();
+            }*/
+
+            m_rigidBody.MovePosition(m_rigidBody.transform.position + movement);
+
+            m_playAreaPreviousY = transform.position.y;
+        }
+
+        public Vector3 getMovementDirection()
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_RealMovement.cs]: Public methods should be above private and protected methods. Also `GetMovementDirection()` (uppercase start)

File: Assets/VRTK/Scripts/VRTK_RealMovement.cs [hunk 22]
```diff
@@ -0,0 +1,350 @@
...
+        {
+            return m_direction;
+        }
+
+        public float getSpeed()
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_RealMovement.cs]: `GetSpeed()`

**Comment by thestonefox** [comment]
Superseded by https://github.com/thestonefox/VRTK/pull/664

**Pull request closed without merging**